### PR TITLE
Wandb log stats

### PR DESCRIPTION
**Pull request by haeggee**
added logging for

- tokens per sec
- consumed tokens
- (correct) model weight norm (global, all params)
- ETA

anything else we need?

### CODE REVIEW

File: megatron/training/training.py [hunk 1]
```diff
@@ -1046,6 +1049,25 @@ def training_log(loss_dict, total_loss_dict, learning_rate, decoupled_learning_r
         throughput = num_floating_point_operations(args, batch_size) / (
             elapsed_time_per_iteration * 10**12 * args.world_size)
 
+        # Calculate tokens per second
+        tokens_per_iteration = args.global_batch_size * args.seq_length
+        tokens_per_sec = tokens_per_iteration / elapsed_time_per_iteration
```
Comment (TJ-Solergibert) [inline megatron/training/training.py:1054]:
```suggestion
        tokens_per_sec = tokens_per_iteration / elapsed_time_per_iteration / torch.distributed.get_world_size()
```
IMO It makes much more sense to divide the throughput with the number of devices, this way we can compare across configs and methods. It doesn't provide any relevant information having `124334` or `8878748676` apart from computing when the job is going to finish but for that we already have `eta`, right?

File: megatron/training/utils.py [hunk 2]
```diff
@@ -97,6 +97,10 @@ def calc_params_l2_norm(model, force_create_fp32_copy=False):
                             if getattr(param, 'main_param_sharded', False):
                                 if param.main_param is not None:
                                     sharded_params_data.append(param.main_param)
+                                else:
+                                    # Fallback to data if `.main_param`
+                                    # is available but is None
+                                    params_data.append(param.data)
```
Comment (TJ-Solergibert) [inline megatron/training/utils.py:103]: Was this raising any issue? Should we reach Megatron-LM?

**Review by TJ-Solergibert — COMMENTED** [review]
Thanks Alex! What do you think about removing `num zeros` from the panel too? And removing `batch-size` if args.rampup_batch_size is None?

File: megatron/training/utils.py (hunk 2 above)
Comment (haeggee) [reply megatron/training/utils.py:103]: it didn't cause any errors but essentially the param_norm can be zero without it (e.g., it happened when i did debugging with the 1B script https://github.com/swiss-ai/Megatron-LM/blob/main/submit-llama3-1b.sh). i don't fully understand the mechanism behind `main_param` or what it's for -- i just saw that in our setup, the attribute exists but is None

File: megatron/training/training.py (hunk 1 above)
Comment (haeggee) [reply megatron/training/training.py:1055]: made the change for tok/sec/gpu. the elapsed time per iter per rank is quite a bit more work, yeah, so maybe we can ignore that for now?

Comment (dhia680) [reply megatron/training/training.py:1055]: yes, not a priority IMO

Comment (TJ-Solergibert) [reply megatron/training/training.py:1055]: Logging per gpu metrics (either the iteration time or loss per dp replica let's say) it's overkilled and will be a completely mess for the Wandb panel. 

Am I the only one who still thinks `tokens_per_sec ` is useless as we already have `tokens_per_sec_per_gpu `? Can we get rid of that?

Comment (TJ-Solergibert) [reply megatron/training/training.py:1055]: You decide. I'm strongly biased towards minimalism, might be my problem too 😅

**Review by TJ-Solergibert — APPROVED** [review]
👌
